### PR TITLE
Provide a way to omit Event stages in audit policy

### PR DESCRIPTION
**Pull request by CaoShuFeng**
This provide a way to omit some stages for each audit policy rule.
    
For example:

    ```
      apiVersion: audit.k8s.io/v1beta1
      kind: Policy
      - level: Metadata
        resources:
           - group: "rbac.authorization.k8s.io"
             resources: ["roles"]
        omitStages:
          - "RequestReceived"
    ```
    
RequestReceived stage will not be emitted to audit backends with previous config.

**Release note**:

```
None
```
#

### CODE REVIEW

**Comment by CaoShuFeng** [comment]
@sttts @tallclair 
I will add unit test for this change soon if my idea is "right".

**Comment by tallclair** [comment]
@CaoShuFeng - What is your motivation for omitting stages? Would the solution proposed in https://github.com/kubernetes/kubernetes/pull/46592#issuecomment-316553044 work for you? I'd like to explore that option before expanding the audit profile API.

**Comment by CaoShuFeng** [comment]
@tallclair Thanks for your review.

I write this pr because of this [expression](https://github.com/kubernetes/kubernetes/issues/48561#issue-241059290):
> We want to omit the RequestReceived stage in GKE. The policy may be the right place to specify that.

It looks like I got it wrong.
What's the meaning behind `The policy may be the right place to specify that.` ?

**Comment by tallclair** [comment]
This isn't wrong, per say. I just wanted to discuss alternatives before putting the burden on the user (i.e. adding it to the policy). If we can accomplish the same goals automatically (the batching proposal) then I think that would be preferable.

**Comment by CaoShuFeng** [comment]
> . I just wanted to discuss alternatives before putting the burden on the user (i.e. adding it to the policy)

Agree.

Thanks for teaching me.
I will leave this pr open for a while. `Maybe` this is usefule.

**Comment by CaoShuFeng** [comment]
Closed according to commants from @tallclair and @ericchiang .
Ref:
https://github.com/kubernetes/kubernetes/issues/48561#issuecomment-318146110
https://github.com/kubernetes/kubernetes/issues/48561#issuecomment-318163694

**Comment by CaoShuFeng** [comment]
Repopened according to @crassirostris 's commants.
Ref:
https://github.com/kubernetes/kubernetes/issues/48561#issuecomment-318432155

This is still WIP, Unit tests will come soon.

**Comment by crassirostris** [comment]
@tallclair suggested earlier that it might be useful to have `omitStages` field in the `Policy` object that will be "ored" with the `omitStages` field in the `PolicyRule`. That's a syntax sugar for cases where a stage has to be excluded in every rule

@sttts @CaoShuFeng WDYT?

**Comment by crassirostris** [comment]
Btw, it should probably should be rebased once https://github.com/kubernetes/kubernetes/issues/49115 is merged and the changes should go to the beta API

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/generated.pb.go [hunk 1]
```diff
@@ -528,6 +528,21 @@ func (m *PolicyRule) MarshalTo(dAtA []byte) (int, error) {
 			i += copy(dAtA[i:], s)
 		}
 	}
+	if len(m.OmitStages) > 0 {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/generated.pb.go]: no need for the if.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/generated.pb.go [hunk 2]
```diff
@@ -723,6 +738,12 @@ func (m *PolicyRule) Size() (n int) {
 			n += 1 + l + sovGenerated(uint64(l))
 		}
 	}
+	if len(m.OmitStages) > 0 {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/generated.pb.go]: no need for the if

File: staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/generated.pb.go (hunk 1 above)
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/v1alpha1/generated.pb.go]: oops, that's generated proto :)

File: staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go [hunk 3]
```diff
@@ -40,21 +40,21 @@ func NewChecker(policy *audit.Policy) Checker {
...
+func FakeChecker(level audit.Level, stage []audit.Stage) Checker {
+	return &fakeChecker{level, stage}
 }
 
 type policyChecker struct {
 	audit.Policy
 }
 
-func (p *policyChecker) Level(attrs authorizer.Attributes) audit.Level {
+func (p *policyChecker) Level(attrs authorizer.Attributes) (audit.Level, []audit.Stage) {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go]: don't the level and stages overlap?

File: staging/src/k8s.io/apiserver/pkg/apis/audit/types.go [hunk 4]
```diff
@@ -200,6 +200,11 @@ type PolicyRule struct {
 	//  "/healthz*" - Log all health checks
 	// +optional
 	NonResourceURLs []string
+
+	// OmitStages specify events generated in which stages will not be emitted to backend
+	// Allowed stages are RequestReceived,ResponseStarted,ResponseComplete,Panic
+	// empty list means every events will be emitted.
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/types.go]: nit: An empty....

File: staging/src/k8s.io/apiserver/pkg/apis/audit/types.go [hunk 5]
```diff
@@ -200,6 +200,11 @@ type PolicyRule struct {
 	//  "/healthz*" - Log all health checks
 	// +optional
 	NonResourceURLs []string
+
+	// OmitStages specify events generated in which stages will not be emitted to backend
+	// Allowed stages are RequestReceived,ResponseStarted,ResponseComplete,Panic
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/types.go]: nit: fullstop and spaces after ,

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go [hunk 6]
```diff
@@ -157,9 +163,10 @@ var _ http.ResponseWriter = &auditResponseWriter{}
 // create immediately an event (for long running requests).
 type auditResponseWriter struct {
 	http.ResponseWriter
-	event *auditinternal.Event
-	once  sync.Once
-	sink  audit.Sink
+	event      *auditinternal.Event
+	once       sync.Once
+	sink       audit.Sink
+	omitStages []auditinternal.Stage
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: please add a comment: omitStages are skipped, i.e. not sent to the backend. But their information is still recorded in the context Event for later stages.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/types.go (hunk 5 above)
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/types.go]: Is it common to relist the constants in those comments?

**Comment by sttts** [comment]
Some nit. Overall lgtm.

**Comment by sttts** [comment]
> @tallclair suggested earlier that it might be useful to have omitStages field in the Policy object that will be "ored" with the omitStages

Sounds good.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/types.go (hunk 5 above)
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/types.go]: > Is it common to relist the constants in those comments?

+1, I don't think it's a great idea

**Comment by CaoShuFeng** [comment]
> @tallclair suggested earlier that it might be useful to have omitStages field in the Policy object that will be "ored" with the omitStages

+1 will add it.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go (hunk 6 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/apis/audit/types.go (hunk 5 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/types.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go (hunk 3 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go]: Renamed to LevelAndStages

File: staging/src/k8s.io/apiserver/pkg/apis/audit/types.go (hunk 4 above)
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/types.go]: Done.

**Comment by CaoShuFeng** [comment]
Unit tests has beed added to it.


omitStages in Policy Object will be added later. And this needs a rebase after #49115

**Comment by CaoShuFeng** [comment]
@crassirostris 
Can you help me start the tests for this pr?
This needs a "/ok-to-test" label.
So that I can see whether it works well.

File: staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go (hunk 3 above)
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/checker.go]: <nit-picking>LevelAndOmitStaged</nit-picking>

**Comment by CaoShuFeng** [comment]
I will rebase this change after https://github.com/kubernetes/kubernetes/pull/48836 gets merged.

**Comment by sttts** [comment]
@CaoShuFeng please rebase and run hack/update-gofmt.sh.

/lgtm
/approve no-issue

File: staging/src/k8s.io/apiserver/pkg/audit/policy/checker_test.go [hunk 7]
```diff
@@ -111,52 +111,73 @@ func TestChecker(t *testing.T) {
 				"/metrics",
 			},
 		},
+		"omit RequestReceived": {
+			Level: audit.LevelRequest,
+			OmitStages: []audit.Stage{
+				audit.Stage("RequestReceived"),
+			},
+		},
+		"only audit painc": {
```
Comment (soltysh) [inline staging/src/k8s.io/apiserver/pkg/audit/policy/checker_test.go]: s/painc/panic

**Review by soltysh — COMMENTED** [review]
Mind fixing the typo while rebasing this PR

**Comment by CaoShuFeng** [comment]
I will rebase this after https://github.com/kubernetes/kubernetes/pull/51119 get merged.
@soltysh

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go [hunk 8]
```diff
@@ -92,65 +116,35 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
...
 		handler.ServeHTTP(respWriter, req)
 	})
 }
 
-// createAuditEventAndAttachToContext is responsible for creating the audit event
-// and attaching it to the appropriate request context. It returns:
-// - context with audit event attached to it
-// - created audit event
-// - error if anything bad happened
-func createAuditEventAndAttachToContext(requestContextMapper request.RequestContextMapper, req *http.Request, policy policy.Checker) (request.Context, *auditinternal.Event, error) {
```
Comment (CaoShuFeng) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: @soltysh sorry for tearing down this function. :(

**Comment by sttts** [comment]
Needs rebase.

**Comment by CaoShuFeng** [comment]
> Needs rebase.

Partially done. This will need another rebase after #51797.
#51797 has gone into the submit queue.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go [hunk 9]
```diff
@@ -42,19 +42,43 @@ func WithAudit(handler http.Handler, requestContextMapper request.RequestContext
 		return handler
 	}
 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		ctx, ev, err := createAuditEventAndAttachToContext(requestContextMapper, req, policy)
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: Can't we return omitStages from `createAuditEventAndAttachToContext`? Am not a fan of this code duplication.

Comment (crassirostris) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: I'm not sure why this function is removed at all. `omitStages` are later obtained from the `policy` and not used until the code that was previously outside of `createAuditEventAndAttachToContext`. @CaoShuFeng Could you please elaborate more?

Comment (CaoShuFeng) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: I have took `createAuditEventAndAttachToContext` back.
And the codes look cleaner for me, thanks.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: much better now

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: lgtm. @crassirostris can you take another look and /lgtm if it is ok?

File: staging/src/k8s.io/apiserver/pkg/apis/audit/types.go [hunk 10]
```diff
@@ -207,6 +207,10 @@ type PolicyRule struct {
 	//  "/healthz*" - Log all health checks
 	// +optional
 	NonResourceURLs []string
+
+	// OmitStages specify events generated in which stages will not be emitted to backend.
+	// An empty list means every events will be emitted.
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/types.go]: I think that might be confusing. What about `An empty list means no restrictions will apply.`?

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go [hunk 11]
```diff
@@ -173,6 +179,8 @@ type auditResponseWriter struct {
 	event *auditinternal.Event
 	once  sync.Once
 	sink  audit.Sink
+	// omitStages are skipped, i.e. not sent to the backend. But their information is still recorded in the context Event for later stages.
```
Comment (crassirostris) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: > omitStages are skipped

Is this comment necessary? `once` and `sink` are not set to the backend also :) Basically, the only explanation why `omitStages` field is there is `b/c it's used later` and this description is not very informative. Maybe remove the comment altogether?

**Review by crassirostris — COMMENTED** [review]
Some nits regarding the comments, otherwise LGTM

File: staging/src/k8s.io/apiserver/pkg/apis/audit/types.go (hunk 10 above)
Comment (CaoShuFeng) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/types.go]: Done.

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go (hunk 11 above)
Comment (CaoShuFeng) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go]: Good for me.
Deleted.

**Review by crassirostris — APPROVED** [review]
LGTM

File: staging/src/k8s.io/apiserver/pkg/apis/audit/types.go [hunk 12]
```diff
@@ -207,6 +207,10 @@ type PolicyRule struct {
 	//  "/healthz*" - Log all health checks
 	// +optional
 	NonResourceURLs []string
+
+	// OmitStages specify events generated in which stages will not be emitted to backend.
+	// An empty list means no restrictions will apply.
```
Comment (tengqm) [inline staging/src/k8s.io/apiserver/pkg/apis/audit/types.go]: should we mark this as optional?

Comment (CaoShuFeng) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/types.go]: Yes, we should.

Comment (CaoShuFeng) [reply staging/src/k8s.io/apiserver/pkg/apis/audit/types.go]: Added here together:
https://github.com/kubernetes/kubernetes/pull/54634

**Pull request merged**